### PR TITLE
Fix lowering code for ES2018 Async Iteration

### PR DESCRIPTION
**Pull request by nhat-nguyen**
Also took this opportunity to do a bit refactor on how we initialize the `generator` pointer inside `ResumeYieldData`. I think it's cleaner if we do everything in the constructor instead of only setting it in `CallAsyncGenerator`.

Related PR: #5834

### CODE REVIEW

File: lib/Runtime/Library/JavascriptGenerator.cpp [hunk 1]
```diff
@@ -325,8 +325,9 @@ namespace Js
 
         AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
 
-        ResumeYieldData yieldData(args[1], RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, args[1], scriptContext, nullptr));
-        resumeNextReturnProcessor->GetGenerator()->CallAsyncGenerator(&yieldData);
+        JavascriptGenerator* generator = resumeNextReturnProcessor->GetGenerator();
+        ResumeYieldData yieldData(args[1], RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, args[1], scriptContext, nullptr), generator);
+        generator->CallAsyncGenerator(&yieldData);
```
Comment (zenparsing) [inline lib/Runtime/Library/JavascriptGenerator.cpp]: Would it be possible to just pass in the `data` and `exceptionObj` to `CallAsyncGenerator`, and let `CallAsyncGenerator` deal with creating the ResumeYieldData structure?

```cpp
void CallAsyncGenerator(Var data, JavascriptExceptionObject* exceptionObj);
```

That way we wouldn't have to deal with the odd situation where (in principle) the `generator` field could be something other than `this`.

Comment (nhat-nguyen) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: That's how it is currently done actually. But I thought it was a bit hard to follow for someone who is not familiar with the code.

I found myself wondering where do we set this `generator` field (only one place yet required me to do a search all of uses of this field because we don't use setter, and it's not when the ResumeYieldData is constructed).

Whereas here, since all fields are const, I can assume they are all set during construction, and simply have to look at where we create new ResumeYieldData.

Granted, I agree that delaying setting it to CallAsyncGenerator makes it a bit cleaner as you suggested - so I guess it's a tradeoff of readability vs. cleanliness. But I still find it weird to skip setting it in the constructor only to do it a function call right after especially when we have all the information at that point.

Comment (nhat-nguyen) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: Yeah I think that is a better way of doing it since we only need the generator in CallAsyncGenerator anyway.

Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: Sorry for not writing this the best way in the first place - as is likely obvious I started by copying the existing Generator code then editing as needed this somewhat obtuse construction is a holdover from that.

Comment (nhat-nguyen) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: No worries, I just updated the signature for CallAsyncGenerator as @zenparsing  suggested. 😄
